Bluetooth: Convert HCI drivers to use a proper Zephyr driver API

jhedberg opened the pull request:
Current status:

- [x] Create a base binding and API definition
- [x] Update the host to support new HCI driver instances
- [x] Update `h4.c`
- [x] Update native controller HCI driver
- [x] Fix `samples/bluetooth/hci_uart_async/` to work with the new driver model
- [x] Fix bsim to work both with native and UART-based controllers (right now CI is failing)
- [x] Figure out what to do with `bt_read_static_addr()`. Taken care of by #72580
- [x] Figure out what to do with `BT_HAS_HCI_VS`, `BT_HCI_VS` and `BT_HCI_VS_EXT` (perhaps DTS properties?). Some preparation in #72580.
- [x] Fix controller unit tests: no devicetree available for `unit_testing` "board" right now: #72871
- [x] Update `ipc.c`
- [x] Update `userchan.c`
- [x] Update `h5.c`
- [x] Update `hci_ambiq.c` (and associated DT binding)
- [x] Remove `hci_b91.c`: #73289
- [x] Update `hci_esp32.c`
- [x] Update `hci_psoc6_bless.c` (and associated DT binding)
- [x] Update `hci_spi_st.c` (and associated DT binding)
- [x] Update `hci_stm32wba.c`
- [x] Update `ipm_stm32wb.c` (and associated DT binding)
- [x] Update `spi.c`
- [x] Update `hci_da1469x.c`
- [x] Update `hci_nxp.c`
- [x] Update `slz_hci.c`
- [x] Update documentation, especially stuff regarding `/chosen` properties
- [x] Make it possible for `nrf52_bsim` to use also UART for HCI
- [x] Update various apps under `samples/` and `tests/` that use the legacy HCI driver API
- [x] Get rid of the Kconfig `choice` section for HCI drivers
- [x] Get rid of `BT_NO_DRIVER`
- [x] Deprecate the old API
- [ ] Lots of testing
  - [x] bsim with virtual driver (hci_driver.c)
  - [x] ipc.c tested in CI
  - [ ] Test on hardware with virtual bus (i.e. combined build)
  - [ ]  Test on hardware with IPC (i.e. split build)
  - [ ]  Test on hardware with UART (i.e. split build)
  - [x]  userchan.c
  - [ ] Test with an OoT driver (e.g. NCS)

Right now the host `hci_core.c` uses `#if DT_HAS_CHOSEN(zephyr_bt_hci)` to separate the new-style HCI drivers from the legacy HCI driver support. The old API has been deprecated, but will only be removed after a deprecation period.

The following drivers depend on "west blobs", so they're not covered by CI, but were manually compile tested:

- `hci_psoc6_bless.c` (`west blobs fetch hal_infineon`)
- `hci_esp32.c` (`west blobs fetch hal_espressif`)
- `hci_stm32wba.c` (`west blobs fetch hal_stm32` doesn't work for me, see https://github.com/zephyrproject-rtos/zephyr/pull/72323#issuecomment-2143443846 )
- `slz_hci.c` (`west blobs fetch hal_silabs`, same problem for me as with stm32, i.e. had to manually download)

What this PR *doesn't* try to tackle:

- Find a solution for and fix driver dependencies on host HCI command sending APIs, which are e.g. not available with HCI_RAW. https://github.com/zephyrproject-rtos/zephyr/blob/d3c0dfbf3699fc98e01909ecf9a2ee834d7f4d41/drivers/bluetooth/hci/apollox_blue.c#L289-L333
- Decide what to do with `BT_HCI_ACL_FLOW_CONTROL` currently referencing specific HCI drivers
- net_buf allocations by drivers. Drivers should own their own memory rather than calling into the host side to do these allocations.
- Making the HCI driver API support system calls. This is related to the above first point. We also can't have the driver make direct calls using the `recv` callback to the host, rather there needs to be another mechanism for the host to fetch data from the driver.

Fixes #51503

jhedberg committed:
Bluetooth: Add devicetree base binding for Bluetooth HCI drivers

Add a base binding for Bluetooth HCI drivers. All HCI drivers should
extend this binding (through an include directive) to create their own
binding files.

Signed-off-by: Johan Hedberg <johan.hedberg@gmail.com>

jhedberg committed:
Bluetooth: include: Add initial HCI driver API definition

Add the initial HCI driver header file and driver API definition. This
is distinct from drivers/bluetooth/hci_driver.h in order to support
legacy (unconverted) drivers for the time being.

Signed-off-by: Johan Hedberg <johan.hedberg@gmail.com>

jhedberg committed:
Bluetooth: host: Add support for new-style HCI drivers

Add support for HCI drivers which use the newly defined HCI driver API.
Since Zephyr (currently) only supports a single HCI driver instance,
boards are expected to indicate the instance using a new devicetree
chosen property `zephyr,bt_hci`.

In order to maintain compatibility with not-yet-converted drivers the
code has been placed behind `#if DT_HAS_CHOSEN(zephyr_bt_hci)`
conditionals.

Signed-off-by: Johan Hedberg <johan.hedberg@gmail.com>

jhedberg committed:
Bluetooth: drivers: Convert H4 (UART) HCI driver to new API

Convert the H4 driver to the new HCI driver API. This includes updating
also any boards that use the driver, i.e. adding the appropriate
devicetree node and chosen property to them.

Signed-off-by: Johan Hedberg <johan.hedberg@gmail.com>

jhedberg committed:
Bluetooth: controller: Update to new HCI driver API

Update the native controller to the new HCI driver API. The devicetree
node is placed under existing `radio` nodes, which seemed like the most
intuitive option.

Signed-off-by: Johan Hedberg <johan.hedberg@gmail.com>

jhedberg commented on dts/arm/nordic/nrf52832.dtsi:
```diff
@@ -90,6 +91,11 @@
 			interrupts = <1 NRF_DEFAULT_IRQ_PRIORITY>;
 			status = "okay";
 			ble-2mbps-supported;
```
One open question is, should properties like these be moved from the `radio` node to the `bt_hci` node?

jhedberg commented on subsys/bluetooth/controller/hci/hci_driver.c:
```diff
@@ -167,6 +180,8 @@ isoal_status_t sink_sdu_emit_hci(const struct isoal_sink             *sink_ctx,
 				 const struct isoal_emitted_sdu_frag *sdu_frag,
 				 const struct isoal_emitted_sdu      *sdu)
 {
+	const struct device *dev = DEVICE_DT_GET(DT_DRV_INST(0));
```
This is not particularly clean, however I didn't immediately see a way to get the `const struct device *` routed to the `sink_sdu_emit_hci()` call. In all the other cases (e.g. the rx threads) I was able to pass the `struct device` as a callback parameter.

jhedberg committed:
Bluetooth: samples: hci_uart_async: Remove bt_recv() dependency

Apps shouldn't try to call bt_recv() since that's not a public API and
is going completely away with the new HCI driver API. Just inject the
buffer directly into the k_fifo which we anyway have access to in this
app.

Signed-off-by: Johan Hedberg <johan.hedberg@gmail.com>

jhedberg committed:
Bluetooth: host: hci_raw: Use existing H4 defines from hci_types.h

Use existing defines instead of redefining our own.

Signed-off-by: Johan Hedberg <johan.hedberg@gmail.com>

jhedberg committed:
Bluetooth: drivers: h5: Convert to new HCI driver API

Convert the H:5 HCI driver to use the new HCI driver API.

Signed-off-by: Johan Hedberg <johan.hedberg@gmail.com>

jhedberg committed:
Bluetooth: drivers: hci: Get rid of Kconfig choice

The drivers should be independent after the move to the new HCI driver
API. Having them as a choice also has unexpected consequences with some
drivers being unexpectedly enabled.

Signed-off-by: Johan Hedberg <johan.hedberg@gmail.com>

jhedberg committed:
boards: nrf52_bsim: Add support for using UART as HCI

Add support for using an UART instead of the native controller. This is
accomplished with a custom DTS overlay for the tests that require it.

Signed-off-by: Johan Hedberg <johan.hedberg@gmail.com>

jhedberg committed:
Bluetooth: drivers: Convert userchan driver to new API

Convert the HCI User Channel driver to use the new HCI driver API.

Signed-off-by: Johan Hedberg <johan.hedberg@gmail.com>

jhedberg committed:
Bluetooth: tests: Convert to new HCI driver API

Convert all Bluetooth tests that define dummy HCI driver instances to use
the new HCI driver API. This requires both a custom DTS binding as well
as an app-specific overlay file.

Signed-off-by: Johan Hedberg <johan.hedberg@gmail.com>

jhedberg committed:
Bluetooth: drivers: userchan: don't fail everything if driver init fails

With the new HCI driver model it's not always critical if a single
instance fails to initialize. This is especially true for many test
applications which provide their own HCI drivers. Instead of causing a
complete failure, simply fail to initialize the driver instance.

Signed-off-by: Johan Hedberg <johan.hedberg@gmail.com>

jhedberg committed:
qemu: Use BT_UART instead of BT_NO_DRIVER to enable serial port

The CONFIG_BT_NO_DRIVER option will be going away. Use the
CONFIG_BT_UART option instead (selected by UART HCI drivers) to
determine whether to enable the qemu serial port for Bluetooth.

Signed-off-by: Johan Hedberg <johan.hedberg@gmail.com>

aaronyegx commented:
@jhedberg May I know what the main issue will be addressed by this driver API converting? It seems the changes are wide so I would like to know more about the background and see what comments I can give for HCI Ambiq codes. Thanks.

jhedberg commented:
> @jhedberg May I know what the main issue will be addressed by this driver API converting? It seems the changes are wide so I would like to know more about the background and see what comments I can give for HCI Ambiq codes. Thanks.

@aaronyegx most of the justifications can be found in the linked issue, i.e. #51503. Additionally, longer term we're considering the possibility of having the Bluetooth host run in user space. The primary way for user space to communicate with kernel space is through the system call interface, and there's a well established way how to do this for Zephyr driver APIs. Note that this aspect is something that's going to happen much later in the future (if at all) - the justifications in #51503 should be sufficient in themselves.

jhedberg committed:
Bluetooth: samples: Fix PAwR samples to match board capabilities

These samples were enabled for an nRF board, even though there is no
upstream support for the feature. Additionally the sample.yaml files
were forcefully disabling any in-tree HCI drivers, making it impossible
to use e.g. qemu with an external HCI controller. Simply remove the nRF
board from the list, as well as the odd Kconfig HCI driver options.

Signed-off-by: Johan Hedberg <johan.hedberg@gmail.com>

jhedberg committed:
Bluetooth: drivers: Convert Ambiq Apollo driver to new API

Convert the Ambiq Apollo HCI driver to the new HCI API.

Signed-off-by: Johan Hedberg <johan.hedberg@gmail.com>

jhedberg commented:
@aaronyegx I've pushed a conversion commit for the Ambiq HCI driver (https://github.com/zephyrproject-rtos/zephyr/pull/72323/commits/f7c34d40791099f6154da7d96bddc8d518c31545). It's a very simplistic one which doesn't e.g. try to add support for multiple instances (i.e. I keep the global variables instead of moving them to the instance structs). Could you take a look? I was able to verify that it builds, but I have no idea if it still works on actual HW.

jhedberg committed:
Bluetooth: Kconfig: Remove BT_NO_DRIVER

There are no actual users in the main tree anymore, so just remove this
option.

Signed-off-by: Johan Hedberg <johan.hedberg@gmail.com>

aaronyegx commented:
> @aaronyegx I've pushed a conversion commit for the Ambiq HCI driver ([4818516](https://github.com/zephyrproject-rtos/zephyr/commit/4818516a60e6bca5156acddf40b7e8b6b7c4852b)). It's a very simplistic one which doesn't e.g. try to add support for multiple instances (i.e. I keep the global variables instead of moving them to the instance structs). Could you take a look? I was able to verify that it builds, but I have no idea if it still works on actual HW.

@jhedberg Thank you. I will test and update the Ambiq EVB test result soon.

jhedberg committed:
Bluetooth: Use device tree to indicate vendor exension support

Introduce a new bt-hci-vs-ext device tree boolean property to indicate
device tree support.

Signed-off-by: Johan Hedberg <johan.hedberg@gmail.com>

jhedberg commented:
> > @aaronyegx I've pushed a conversion commit for the Ambiq HCI driver ([4818516](https://github.com/zephyrproject-rtos/zephyr/commit/4818516a60e6bca5156acddf40b7e8b6b7c4852b)). It's a very simplistic one which doesn't e.g. try to add support for multiple instances (i.e. I keep the global variables instead of moving them to the instance structs). Could you take a look? I was able to verify that it builds, but I have no idea if it still works on actual HW.
> 
> @jhedberg Thank you. I will test and update the Ambiq EVB test result soon.

@aaronyegx since #72609 was merged, I went ahead and updated this PR to support ambiq3 platforms as well: https://github.com/zephyrproject-rtos/zephyr/pull/72323/commits/15d8dd9a411a8d299397e2905a226eb2b7e694c7

Feel free to test :)

jhedberg committed:
Bluetooth: drivers: Convert IPC driver to new API

Convert the ipc.c HCI driver to the new HCI driver API.

Signed-off-by: Johan Hedberg <johan.hedberg@gmail.com>

jhedberg commented on dts/bindings/bluetooth/zephyr,bt-hci-ipc.yaml:
```diff
@@ -0,0 +1,17 @@
+    default: "BT_HCI_BUS_IPM"
+  bt-hci-quirks:
+    default: ["BT_HCI_QUIRK_NO_AUTO_DLE"]
+  bt-hci-ipc-name:
+    type: string
+    default: "nrf_bt_hci"
+    description: IPC endpoint name
```
Note: the default "nrf_bt_hci" string is probably not the best one, since there's nothing nrf-specific about this. What's important is that this matches what the controller side uses, i.e. in most cases `samples/bluetooth/hci_ipc`:
https://github.com/zephyrproject-rtos/zephyr/blob/53ddff639562ef68dc0a6f62b82f7505c75ebdce/samples/bluetooth/hci_ipc/src/main.c#L369-L370

On the app-side it might be better to have this as an app-specific Kconfig option.

@kapi-no FYI

jhedberg replied:
> It's based on openamp, right? Maybe that can replace the nRF part somehow? Or am I misunderstanding this comment?

I just meant that we could remove "nrf_" from the string. Unltimately all that matters is that both sides of the IPC use the same string since otherwise they will not discover each other.

jhedberg committed:
Bluetooth: drivers: Convert psoc6_bless driver to new API

Convert the hci_psoc6_bless.c HCI driver to the new HCI driver API.

Signed-off-by: Johan Hedberg <johan.hedberg@gmail.com>

jhedberg commented:
@ifyall would be good if someone checked that the psoc6_bless driver still works with this PR (see https://github.com/zephyrproject-rtos/zephyr/pull/72323/commits/19c6f37235e4807e47aeff66b22abb45bda0068e). I was able to compile test it but don't have the HW to do runtime testing.

jhedberg committed:
Bluetooth: drivers: Convert ESP32 HCI driver to new API

Convert the hci_esp32.c HCI driver to the new HCI driver API.

Signed-off-by: Johan Hedberg <johan.hedberg@gmail.com>

jhedberg commented:
@sylvioalves would be good if someone did some runtime testing of the ESP32 HCI driver changes in this PR (https://github.com/zephyrproject-rtos/zephyr/pull/72323/commits/0cb7f9977a285fbb52af7474cf61a70ef6fb49aa). It seems to build fine, but I haven't been able to test it on actual hardware.

jhedberg committed:
Bluetooth: drivers: Convert SPI drivers to use new HCI API

Convert the spi.c and hci_spi_st.c drivers to use the new HCI driver API.
Both drivers are converted in one go since one derives from the other's
devicetree binding.

Signed-off-by: Johan Hedberg <johan.hedberg@gmail.com>

jhedberg committed:
Bluetooth: drivers: Convert ST STM32WBA driver to new API

Convert the hci_stm32wba.c driver to the new HCI API. Unlike in most cases,
the devicetree node is already enabled on the SoC level (rather than board
level). This is in order to mirror how the Kconfig option was originally
enabled, i.e. on the SoC level.

Signed-off-by: Johan Hedberg <johan.hedberg@gmail.com>

jhedberg commented:
@erwango could you or someone else with the HW test that the `hci_stm32wba.c HCI driver still works as expected with this PR? (https://github.com/zephyrproject-rtos/zephyr/pull/72323/commits/8f84c50ceb9b07b8c56fb3e28a89eb7333603059). Another problem I had with this one was the blobs support, at least for me `west blobs fetch hal_stm32` did not fetch the binaries and I had to manually download them instead. It looks like "west blobs" should be currently supported by hal_stm32, however for some reason it didn't work for me. Could you check if this is some local thing or maybe something is wrong with the stm32 HAL definition? Blobs from other modules (like espresso) seem to work fine for me.

jhedberg committed:
Bluetooth: drivers: Convert STM32 IPM driver to new API

Convert the ipm_stm32wb.c HCI driver to the new HCI driver API.

Signed-off-by: Johan Hedberg <johan.hedberg@gmail.com>

jhedberg commented:
@erwango and please also take a look at the conversion of the STM32 IPM HCI driver: https://github.com/zephyrproject-rtos/zephyr/pull/72323/commits/5ce503f401ba47b25a4f7b475d57136f50c9cb47

jhedberg committed:
Bluetooth: drivers: Convert DA1469X HCI driver to new API

Convert the Renesas DA1469X HCI driver to the new HCI driver API.

Signed-off-by: Johan Hedberg <johan.hedberg@gmail.com>

jhedberg committed:
Bluetooth: drivers: Convert NXP HCI driver to new API

Convert the hci_nxp.c HCI driver to use the new HCI driver API. Also move
the driver binding under dts/bindings/bluetooth, like all other HCI driver
bindings.

Signed-off-by: Johan Hedberg <johan.hedberg@gmail.com>

jhedberg commented:
@yeaissa @dleach02 it'd be good if the NXP driver conversion in this PR was tested (commit https://github.com/zephyrproject-rtos/zephyr/pull/72323/commits/3d58d23385e413d9b417b8e40b18568693837221). It seems to build fine at least.

jhedberg committed:
Bluetooth: drivers: Convert Silabs HCI driver to new API

Convert the slz_hci.c HCI driver to use the new HCI driver API. This also
fixes the HCI bus type to correctly indicate VIRTUAL instead of UART.

Signed-off-by: Johan Hedberg <johan.hedberg@gmail.com>

jhedberg commented:
@fkokosinski are you able to test the `slz_hci.c` changes in this PR (commit https://github.com/zephyrproject-rtos/zephyr/pull/72323/commits/d964da8ed32638652213671ac13f4f58b1017116)?

jhedberg commented:
> @fkokosinski are you able to test the `slz_hci.c` changes in this PR (commit [d964da8](https://github.com/zephyrproject-rtos/zephyr/commit/d964da8ed32638652213671ac13f4f58b1017116))?

Also, can you confirm that the controller really sits behind an UART like it claims:
https://github.com/zephyrproject-rtos/zephyr/blob/fe03e0c4bb3f8a658efe69b04c430d1a15474c6c/drivers/bluetooth/hci/slz_hci.c#L198

It seems a bit odd that it'd be requiring so many binary blobs in that case.

jhedberg commented:
I now promoted this to a normal PR since all HCI drivers have been converted. I still added the DNM label, since there are documentation updates which are still missing. We also need to decide whether to keep the `bt_hci_driver_register()` API or remove it. With this PR there are no more users of it in the upstream tree.

yeaissa commented on drivers/bluetooth/hci/hci_nxp.c:
```diff
@@ -346,13 +355,16 @@ static int bt_nxp_open(void)
 		}
 #endif
+
+		hci->recv = recv;
```
@jhedberg Could you please move this to be called before` bt_nxp_set_calibration_data()`? This function is sending HCI messages so hci_rx_cb() gets called and uses this `recv` which still points to `NULL`.

Beside this, I did a sanity check on `rd_rw612_bga` and it worked fine with this fix. 

Thanks for the effort.

jhedberg replied:
@yeaissa thanks for testing this. The reason I didn't catch this particular issue, is that it's not really valid for the driver's `open()` callback to try to send HCI data, at least not using `bt_hci_cmd_send*()` APIs. From the host's perspective open() is what's used to "open" the HCI transport, and it will assume that the transport is not available until that driver callback has returned success. It's by chance that we don't have any checks or host-side state machine initializations tied to this, and so the existing code just works. The right place to do HCI command sending is instead the `setup()` callback - that's really the purpose of it, i.e. anything the HCI driver might need to do before the host sends its first HCI packet.

yeaissa replied:
@jhedberg, I understand your point and it makes sense. I'll update in a separate PR. Thanks.
I guess for now you can just move it in the meantime.

jhedberg replied:
@yeaissa thanks. I incorporated your commit from #73711 into this one (it's the first commit here). I'll rebase once #73711 has been merged.

yeaissa replied:
Thanks @jhedberg, It's merged now.

fkokosinski commented:
> @fkokosinski are you able to test the `slz_hci.c` changes in this PR (commit [121c14f](https://github.com/zephyrproject-rtos/zephyr/commit/121c14fcc383c6f38ae92805747efbba426032f1))?

Hey, unfortunately I don't have the HW at hand. I'm no longer maintaining SiLabs platforms in Zephyr, but maybe someone from SiLabs could help you with this?

aaronyegx reviewed (dismissed):
The Ambiq file changes look good to me.

npal-cy commented:
hi @jhedberg,
the cy8cproto_062_4343w board does not work ('west build -b cy8cproto_062_4343w samples\bluetooth\peripheral_dis')

i do not see CONFIG_DT_HAS_ZEPHYR_BT_HCI_UART_ENABLED in .config,

in boards\infineon\cy8cproto_062_4343w\cy8cproto_062_4343w.dts we use compatible  `infineon,cyw43xxx-bt-hci`, which actually is extension of H4. 
```
	bt-hci {
		status = "okay";
		compatible = "infineon,cyw43xxx-bt-hci";
		bt-reg-on-gpios = <&gpio_prt3 4 (GPIO_ACTIVE_HIGH)>;

		/* Configuration UART speeds for firmware download (fw-download-speed) and
		 * HCI operation (hci-operation-speed).
		 * If hci-operation-speed or fw-download-speed are not defined in bt-hci{...}
		 * node, cyw43xx driver will use bus/current-speed as default speed.
		 */
		fw-download-speed = <3000000>;
	};

```
Should we update infineon,cyw43xxx-bt-hci binding or board dtsi to addition enable zephyr,bt-hci-uart as well.

Regards,
Nazar

jhedberg commented:
> Should we update infineon,cyw43xxx-bt-hci binding or board dtsi to addition enable zephyr,bt-hci-uart as well.

@npal-cy thanks for discovering this issue. It seems I overlooked the various extensions to the standard drivers, like `h4.c` (basically I thought the extensions wouldn't be drivers in themselves, having their own DT_DRV_COMPAT, etc. The `h4.c` driver gets enabled by having a `zephyr,bt-hci-uart` DTS node, so I think that can't be avoided. The way I'd structure this, is to place the `zephyr,bt-hci-uart` node as a child of the actual UART node (like I've done in other places in this PR), and then the `infineon,cyw43xxx-bt-hci` node would be a child of that one. That way you could just reference `DT_PARENT()` from the cyw43xxx driver instead of relying on `DT_CHOSEN(zephyr_bt_uart)` (a property which should really go away as part of this PR). Does that sound alright?

jhedberg commented:
@npal-cy I've now pushed an update to the H4 commit in this PR which does the necessary change. It turns out I had already done the same restructuring for the giga_r1 board, but not for the Infineon board.

jerome-pouiller commented on drivers/bluetooth/hci/slz_hci.c:
Not related to this PR, but any idea why this driver does not follow naming of the other driver? Should not it be named `hci_silabs.c` ?

jhedberg replied:
> Not related to this PR, but any idea why this driver does not follow naming of the other driver? Should not it be named `hci_silabs.c` ?

You're right that it'd be better if there was some consistency with the naming. The drivers are themselves inside a directory called `hci`, so having `hci` also in the file name is a bit redundant. Also, the first drivers that appeared here (e.g. `h4.c`) don't have any prefix. The one thing that does cause some confusion even to me, are the various extension drivers which implement the `bt_hci_transport_setup()` function (declared as `__weak` by e.g. `h4.c` and `ipc.c`). It's not easy to distinguish those just by looking at the filenames.

jerome-pouiller commented on drivers/bluetooth/hci/Kconfig:
```diff
@@ -76,6 +76,8 @@ config BT_STM32WBA
 	bool "Silicon Labs Bluetooth interface"
+	default y
+	depends on DT_HAS_SILABS_BT_HCI_ENABLED
 	depends on SOC_SERIES_EFR32BG22 || SOC_SERIES_EFR32MG24 || SOC_SERIES_EFR32BG27
```
If I understand right, `DT_HAS_SILABS_BT_HCI_ENABLED` should replace `SOC_SERIES_EFR32BG22 || SOC_SERIES_EFR32MG24 || SOC_SERIES_EFR32BG27`, shouldn't?

jhedberg replied:
> If I understand right, `DT_HAS_SILABS_BT_HCI_ENABLED` should replace `SOC_SERIES_EFR32BG22 || SOC_SERIES_EFR32MG24 || SOC_SERIES_EFR32BG27`, shouldn't?

Yes, that's a good point. Those should indeed be removed, thanks!

jerome-pouiller commented on dts/arm/silabs/efr32fg1p.dtsi:
```diff
@@ -173,6 +173,11 @@
+	bt_hci_silabs: bt_hci_silabs {
+		compatible = "silabs,bt-hci";
+		status = "okay";
+	};
```
As mentioned by chat, none of the `efr32fg*` support Bluetooth.

jhedberg replied:
> As mentioned by chat, none of the `efr32fg*` support Bluetooth.

I've tried to keep this PR 1:1 with the existing enablement of the drivers through Kconfig. So if a driver would get enabled before this PR for a certain platform, then it would get enabled also after this PR. I could do this kind of fixes also as part of this PR, but I have a feeling it'd be better to do it separately. It's possible (maybe even likely) that this PR will only go in after the next release, however these incorrect indications of Bluetooth support should probably be fixed before the release.

jerome-pouiller replied:
I believe the condition `depends on SOC_SERIES_EFR32BG22 || SOC_SERIES_EFR32MG24 || SOC_SERIES_EFR32BG27` prevented instantiation of HCI driver on wrong boards.

Maybe we should keep this PR as-is for now and prepare a new PR to remove this weird condition.

jerome-pouiller commented on dts/arm/silabs/efr32xg13p.dtsi:
```diff
@@ -181,6 +181,11 @@
+	bt_hci_silabs: bt_hci_silabs {
+		compatible = "silabs,bt-hci";
+		status = "disabled";
+	};
```
As mentioned by chat, in the `efr32xg13` series, only `efr32mg13` and `efr32bg13` support bluetooth.

jerome-pouiller replied:
I believe this declaration should move to `dts/arm/silabs/efr32bg13p632f512gm48.dtsi`.

jerome-pouiller commented on dts/bindings/bluetooth/silabs,bt-hci.yaml:
```diff
@@ -0,0 +1,13 @@
+  bt-hci-name:
+    default: "sl:bt"
+  bt-hci-bus:
+    default: "BT_HCI_BUS_UART"
```
Just to be curious, this driver does not use UART. What is the meaning of `BT_HCI_BUS_UART`? (this value is not a new thing, it was already there)

jhedberg replied:
> Just to be curious, this driver does not use UART. What is the meaning of `BT_HCI_BUS_UART`? (this value is not a new thing, it was already there)

I asked the same question here: https://github.com/zephyrproject-rtos/zephyr/pull/72323#issuecomment-2143776113

It looks like it was wrong from the start. It should likely be IPM (if the controller is on another core) or VIRTUAL (if the controller is on the same core).

jerome-pouiller replied:
So, it should be `VIRTUAL`.

jhedberg replied:
Regarding where these come from, they're purely informational and map to the BTSNoop protocol that's supported by the Linux (BlueZ) btmon tool:
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/net/bluetooth/hci.h#n58
They map to the "Bus" parameter of the "New Index" BTSnoop packet:
https://git.kernel.org/pub/scm/bluetooth/bluez.git/tree/doc/btsnoop.txt

More info on how to use this with Zephyr: https://docs.zephyrproject.org/latest/connectivity/bluetooth/bluetooth-tools.html#embedded-hci-tracing

npal-cy commented on boards/infineon/cy8cproto_062_4343w/cy8cproto_062_4343w.dts:
```diff
@@ -58,17 +58,22 @@ uart2: &scb2 {
+	bt_hci_uart: bt_hci_uart {
+		compatible = "zephyr,bt-hci-uart";
+
+		bt-hci {
```
let's rename `bt-hci` to `murata-1dx`

jhedberg replied:
> let's rename `bt-hci` to `murata-1dx`

sure

npal-cy commented:
> @ifyall would be good if someone checked that the psoc6_bless driver still works with this PR (see [e361ac4](https://github.com/zephyrproject-rtos/zephyr/commit/e361ac4df349368c3f24c90f1de4f59053d7d8d7)). I was able to compile test it but don't have the HW to do runtime testing.

> @npal-cy I've now pushed an update to the H4 commit in this PR which does the necessary change. It turns out I had already done the same restructuring for the giga_r1 board, but not for the Infineon board.

Hi @jhedberg , 
for now there are two Infineon's boards with enabled BLE:
 1. cy8cproto_062_4343w (use H4 extension ) 
 2. cy8cproto_063_ble  (use bless hci driver) 
 
 I checked your changes on both kits and it works.

jhedberg committed:
Bluetooth: deprecate old HCI driver APIs

Deprecate the bt_recv() and bt_hci_driver_register() APIs. There are no
more users in the tree itself, but we should let any out of tree users
migrate before removing these APIs completely.

Signed-off-by: Johan Hedberg <johan.hedberg@silabs.com>

rugeGerritsen commented on include/zephyr/drivers/bluetooth.h:
```diff
@@ -0,0 +1,253 @@
+ * is safe to start calling the send() handler.
+ *
+ * @param dev  HCI device
+ * @param recv Callback to receive data from the controller
```
Can we document in which execution context `bt_hci_recv_t recv` is to be called?

jhedberg replied:
> Can we document in which execution context `bt_hci_recv_t recv` is to be called?

Yes, added.

rugeGerritsen commented on include/zephyr/drivers/bluetooth.h:
```diff
@@ -0,0 +1,253 @@
+/**
+ * @brief Send HCI buffer to controller.
+ *
+ * Send an HCI command or ACL data to the controller. The exact
```
This API can also be used for ISO, so it would be nice to make it a bit more generic
```suggestion
 * Send an HCI packet to the controller. The exact
```

rugeGerritsen replied:
Is the buffer H4 formatted? It would be nice to document that

jhedberg replied:
Done

jhedberg commented:
Just an update to the people following (and potentially testing) this: it was decided in yesterday's Bluetooth meeting that this will be targeted for the Zephyr 3.7 release. The old API will still remain, but be marked as deprecated.

rugeGerritsen commented on include/zephyr/drivers/bluetooth.h:
```diff
@@ -0,0 +1,253 @@
+ *
+ * @return 0 on success or negative error number on failure.
+ */
+static inline int bt_hci_setup(const struct device *dev, struct bt_hci_setup_params *params)
```
How do we envision that this API is extended for future use cases?
Right now it is only used to set public address, but later it may be used to configure power mode or something else.
Will there be cases where some parameters are used by some vendors only?

As the API definition and driver code are located in the same place we know that these will always be in sync, so it may not be a big issue. If there is something vendor specific we can hide it behind KConfig options

jhedberg replied:
> How do we envision that this API is extended for future use cases?
> Right now it is only used to set public address, but later it may be used to configure power mode or something else.
> Will there be cases where some parameters are used by some vendors only?

All good questions. The parameter is in the form of a struct in order to provide some more flexibility and easier extension in the future. I don't see how this could contain vendor-specific extensions though, since our host-side code is (at least right now) rather vendor agnostic.

In general, discussing future of this API is something we should certainly do, but I feel that it's out of scope of this PR. The size of this PR is already quite large as-is, so I've tried to keep the semantics as close to 1:1 to the old API as possible, and leave any further API redesign for later.

jhedberg committed:
Bluetooth: doc: Update documentation for the new HCI API

Update references to devicetree chosen properties and how devicetree nodes
should be defined in order to enable the new HCI drivers.

Signed-off-by: Johan Hedberg <johan.hedberg@silabs.com>

jhedberg committed:
doc: migration-guide-3.7: Add information about new HCI API

Document the changes coming with the new HCI driver API, and provide
information for how to migrate existing drivers.

Signed-off-by: Johan Hedberg <johan.hedberg@silabs.com>

rugeGerritsen commented on include/zephyr/drivers/bluetooth.h:
```diff
@@ -0,0 +1,253 @@
+	int (*close)(const struct device *dev);
+	int (*send)(const struct device *dev, struct net_buf *buf);
+#if defined(CONFIG_BT_HCI_SETUP)
+	int (*setup)(const struct device *dev,
```
Minor suggestion: Move the definition of `setup` before `open` to make it clear that `setup` will always be called before `open`.

jhedberg replied:
> Minor suggestion: Move the definition of `setup` before `open` to make it clear that `setup` will always be called before `open`.

It's actually the other way around. `setup()` is intended for driver-related HCI commands (often vendor ones) which should be sent before the host starts sending any HCI commands. The HCI transport is not considered to be open before `open()` has returned successfully, which is why the separate `setup()` callback exists.

I think what causes confusion is that some drivers (`h4.c` and `ipc.c` in particular) have chosen `bt_hci_transport_setup()` as the name for a `__weak` function that can be used for low-level initialization of the transport (not related to actual HCI communication).

rugeGerritsen commented on include/zephyr/drivers/bluetooth.h:
```diff
@@ -0,0 +1,253 @@
+ *
+ * @return 0 on success, negative error value on failure
+ */
+int bt_hci_transport_setup(const struct device *dev);
```
Is there a reason for leaving the transport setup and teardown out of the driver API? It looks like these will always used together with `open` and `close`. Maybe we should utilize the setup API to pass in the HCI transport device to be used in `open` and `close`?

jhedberg replied:
This is really just something inherited from the original `include/drivers/bluetooth/hci_driver.h`. This API should probably be moved somewhere else since it's really an extension that some drivers have chosen but it's not really part of the generic HCI driver API (that the host or `hci_raw.c` see).

jhedberg commented:
@rugeGerritsen just as a general note, the documentation for many of the `drivers/bluetooth.h` APIs have been copied from the original `drivers/bluetooth/hci_driver.h`, i.e. the original documentation likely suffers from the same issues. I will address your comments for the new APIs, but since we're deprecating the old API I think I'll leave that untouched.

Thalley commented on include/zephyr/drivers/bluetooth.h:
```diff
@@ -0,0 +1,257 @@
+ * @param dev HCI device
+ * @param buf Buffer containing data to be sent to the controller.
+ *
+ * @return 0 on success or negative error number on failure.
```
Would prefer if each possible value was documented via `@retval`, alternatively we should consider explicitly writing that it's an `errno` value in case of error. 

This applies to all new API functions added here

Thalley commented on include/zephyr/drivers/bluetooth.h:
```diff
@@ -0,0 +1,257 @@
+ * event code and the total length of the parameters. Upon successful return
+ * the buffer is ready to have the parameters encoded into it.
+ *
+ * @param evt        Event OpCode.
```
Maybe add a reference to the possible opcodes. I assume it's the ones defines by `hci_types.h`, right?

jhedberg replied:
Well, it's really the HCI specification that defines these, e.g. hci_types.h might be missing some.

Thalley replied:
Ah, good point. In any case, I think referring to the define values are helpful for lookups, as we have for most other spec-defined values

Thalley commented on include/zephyr/drivers/bluetooth.h:
```diff
@@ -0,0 +1,257 @@
+ * length of the parameters. Upon successful return the buffer is ready to have
+ * the parameters encoded into it.
+ *
+ * @param op         Command OpCode.
```
Ditto with a reference to the possible opcode values

jhedberg replied:
I added "HCI", in case that makes it any clearer. Btw, there's are exact copies of what's in `include/zephyr/drivers/bluetooth/hci_driver.h`. They're not really part of the HCI driver API, but I had to drag them along to not create scope creep in this PR (we really need to rethink and redesign this aspect of the HCI driver API).

Thalley commented on include/zephyr/drivers/bluetooth.h:
```diff
@@ -0,0 +1,257 @@
+ * code. Upon successful return the buffer is ready to have the parameters
+ * encoded into it.
+ *
+ * @param op         Command OpCode.
```
Ditto

Thalley commented on subsys/bluetooth/controller/hci/hci_driver.c:
```diff
@@ -66,6 +66,15 @@
 
+#define DT_DRV_COMPAT zephyr_bt_hci_ll_sw_split
+
+struct hci_driver_data {
+	bt_hci_recv_t recv;
+};
+
+struct hci_driver_config {
+};
+
```
Can you add a in-code comment why we have a struct with a single field and another struct with no fields?

jhedberg replied:
> Can you add a in-code comment why we have a struct with a single field and another struct with no fields?

The second struct should be removed. I think it's a leftover from when I initially had something in it. The Zephyr driver & device model is such that you can have both read-write and read-only data associated with a `struct device`. The read-only data is found behind the `dev->config` pointer and the read-write behind the `dev->data` pointer. These structs were named accordingly.

Thalley commented on tests/bluetooth/bluetooth/prj.conf:
```diff
@@ -1,6 +1,6 @@
 CONFIG_BT=y
 CONFIG_BT_CTLR=n
-CONFIG_BT_NO_DRIVER=y
+CONFIG_BT_H4=n
```
It may not be obvious that this effectively disables all BT drivers. Perhaps a comment?

jhedberg replied:
Well, it disables two drivers, not all of them. However, these are sufficient considering the board targets that the bluetooth unit tests are built against.

Thalley replied:
> Well, it disables two drivers, not all of them. 

It's all of them for this specific project, isn't it?

Thalley commented on tests/bluetooth/hci_prop_evt/prj.conf:
```diff
@@ -3,7 +3,7 @@ CONFIG_ZTEST=y
 CONFIG_BT=y
 CONFIG_BT_CTLR=n
-CONFIG_BT_NO_DRIVER=y
+CONFIG_BT_H4=n
```
Given how many places you'd had to change `CONFIG_BT_NO_DRIVER=y` to `CONFIG_BT_H4=n` makes me wonder whether we should have kept `CONFIG_BT_NO_DRIVER`

jhedberg replied:
> Given how many places you'd had to change `CONFIG_BT_NO_DRIVER=y` to `CONFIG_BT_H4=n` makes me wonder whether we should have kept `CONFIG_BT_NO_DRIVER`

`BT_NO_DRIVER` relates to the fact that previously all drivers were behind a `choice`, meaning selecting one would deselect all others. As such it doesn't make sense to reintroduce it. With the new model, disabling these is primarily related to avoiding building in unused code. You can have as many drivers as you like, but the current host and hci_raw implementations only support a single instance.

Thalley replied:
My point was that `CONFIG_BT_NO_DRIVER=y` was more obvious than `CONFIG_BT_H4=n` in terms of what we want to achieve here. 

If we keep it like this, perhaps we can simply add a comment stating that we are disabling the only driver that is enabled by default?

Thalley reviewed (changes requested):
A few minor comments and questions

aescolar reviewed (commented):
A couple of minors after a quickish pass.

jhedberg commented on drivers/bluetooth/hci/Kconfig:
> Is there any chance/way that multiple of the DT values can get enabled?
> 
> i.e. can both `DT_HAS_ZEPHYR_BT_HCI_UART_ENABLED` and `DT_HAS_ZEPHYR_BT_HCI_3WIRE_UART_ENABLED` get enabled, which then would enable both Bluetooth drivers?

Yes, you can enable as many drivers as you like, and if a driver supports multiple instances then you can have that as well. The limitation comes from the API user perspective, i.e. the Bluetooth Host stack in most cases, where multiple instances isn't supported yet.

jhedberg commented on doc/releases/migration-guide-3.7.rst:
```diff
@@ -227,6 +227,17 @@ Analog-to-Digital Converter (ADC)
 =============
 
+ * A completely new HCI driver API was introduced and the old one deprecated. The new API
+   follows the "usual" driver API design in Zephyr, with devicetree nodes, etc. The host now
```
> If the "usual" (recommended?) driver API design is documented anywhere, we could add a reference to it here

Maybe I should use some other word. There is just one Zephyr driver & device model - that's what I'm referring to. I didn't immediately find any high-level page on docs.zephyrproject.org that could act as a good reference to this.

jhedberg replied:
I'll reword this to "follows the normal Zephyr driver model"

jhedberg commented on boards/st/steval_stwinbx1/steval_stwinbx1.dts:
```diff
@@ -32,6 +32,8 @@
 		zephyr,sram = &sram0;
 		zephyr,flash = &flash0;
 		zephyr,code-partition = &slot0_partition;
+
+		zephyr,bt-hci = &hci_spi;
```
> Or was this empty line intentional?

The chosen properties in this file were already split in some groups using empty lines. E.g. memory-related properties were in their own group, so I thought it's consistent to have the new HCI property separate too (as it's in a category of its own).

jhedberg commented on dts/arm/nordic/nrf52833.dtsi:
```diff
@@ -9,6 +9,7 @@
 
 / {
 	chosen {
+		zephyr,bt-hci = &bt_hci_controller;
```
Sure, I'll do this:
```diff
diff --git a/dts/arm/nordic/nrf54l15_cpuapp.dtsi b/dts/arm/nordic/nrf54l15_cpuapp.dtsi
index 6a4f5fcae23..f2264079e8d 100644
--- a/dts/arm/nordic/nrf54l15_cpuapp.dtsi
+++ b/dts/arm/nordic/nrf54l15_cpuapp.dtsi
@@ -17,6 +17,10 @@ cpuflpr_vevif: &cpuflpr_vevif_remote {};
 /delete-node/ &cpuflpr_clic;
 
 / {
+	chosen {
+		zephyr,bt-hci = &bt_hci_controller;
+	};
+
 	soc {
 		compatible = "simple-bus";
 		interrupt-parent = <&cpuapp_nvic>;
@@ -29,6 +33,10 @@ cpuflpr_vevif: &cpuflpr_vevif_remote {};
 	};
 };
 
+&bt_hci_controller {
+	status = "okay";
+};
+
 &cpuapp_ppb {
 	compatible = "simple-bus";
 	ranges;
diff --git a/dts/common/nordic/nrf54l15.dtsi b/dts/common/nordic/nrf54l15.dtsi
index 0b8eb1c549a..8896b94f874 100644
--- a/dts/common/nordic/nrf54l15.dtsi
+++ b/dts/common/nordic/nrf54l15.dtsi
@@ -264,6 +264,11 @@
 					compatible = "nordic,nrf-ieee802154";
 					status = "disabled";
 				};
+
+				bt_hci_controller: bt_hci_controller {
+					compatible = "zephyr,bt-hci-ll-sw-split";
+					status = "disabled";
+				};
 			};
 
 			dppic20: dppic@c2000 {
```

jhedberg commented on doc/releases/migration-guide-3.7.rst:
```diff
@@ -227,6 +227,17 @@ Analog-to-Digital Converter (ADC)
 Bluetooth HCI
 =============
 
+ * A completely new HCI driver API was introduced and the old one deprecated. The new API
```
Sure, thanks

jhedberg commented on doc/build/dts/api/api.rst:
```diff
@@ -386,8 +386,8 @@ device.
-   * - zephyr,bt-uart
-     - Sets UART device used by Bluetooth
+   * - zephyr,bt-hci
+     - Selects the HCI device to used by the Bluetooth host stack
```
Done

Thalley commented on include/zephyr/drivers/bluetooth.h:
```diff
@@ -0,0 +1,257 @@
+/** @file
+ *  @brief Bluetooth HCI driver API.
+ *
+ *  Copyright (c) 2024 Johan Hedberg
+ *
+ *  SPDX-License-Identifier: Apache-2.0
+ */
```
It is not clear to me whether this API is experimental, stable or unstable. 
Suggest to add a version to it as per https://docs.zephyrproject.org/latest/develop/api/overview.html#api-overview to indicate it's mature-ness

jhedberg replied:
> It is not clear to me whether this API is experimental, stable or unstable. Suggest to add a version to it as per https://docs.zephyrproject.org/latest/develop/api/overview.html#api-overview to indicate it's mature-ness

@carlescufi can you help out here? I think probably "unstable" is the right categorization. It can't be experimental at least, since we couldn't deprecate the old API otherwise.

carlescufi replied:
Definitely "unstable", correct, no other definition fits this specific usecase. This API is new but already has multiple implementations and is the successor to an established one.

sylvioalves commented:
@jhedberg Working good on esp32 devices.

jhedberg commented:
@jerome-pouiller @Thalley could you take another look? Due to the size of this PR it'd be better to get it in sooner rather than later, especially since the feature freeze is coming this week. Since there doesn't seem to be any fundamental objections to the architectural/API change introduced here, any follow-up fixes & cleanups could happen after the feature freeze as well. There are a couple of other open HCI driver PRs which conflict with this one, which also target the next release, so it'd be good to start merging these sooner rather than later so we get the conflicts resolved and everything merged on time.

Thalley commented:
> @jerome-pouiller @Thalley could you take another look? Due to the size of this PR it'd be better to get it in sooner rather than later, especially since the feature freeze is coming this week. Since there doesn't seem to be any fundamental objections to the architectural/API change introduced here, any follow-up fixes & cleanups could happen after the feature freeze as well. There are a couple of other open HCI driver PRs which conflict with this one, which also target the next release, so it'd be good to start merging these sooner rather than later so we get the conflicts resolved and everything merged on time.

Indeed, and the number of commits easily hide comments on Github (...). 

I'd be OK with merging as is, but I would have to have at least https://github.com/zephyrproject-rtos/zephyr/pull/72323#discussion_r1633432927 resolved ASAP so that we can do things properly (which I believe was the main point of this change in the first place ;) )

carlescufi commented:
@rugeGerritsen @aescolar @yeaissa @erwango @npal-cy @alwa-nordic please take a look ASAP, we would like to include this in Zephyr 3.7.0, which means merging it before this Friday

jerome-pouiller reviewed (commented):
Beside my comment about `efr32xg13`. This PR is fine for me.

rugeGerritsen commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -4280,29 +4348,6 @@ int bt_disable(void)
 	/* Clear BT_DEV_READY before disabling HCI link */
 	atomic_clear_bit(bt_dev.flags, BT_DEV_READY);
 
-	err = bt_dev.drv->close();
```
Do you remember the reason for reordering statements in `bt_disable()`?

jhedberg replied:
> Do you remember the reason for reordering statements in `bt_disable()`?

Yes. There were actions in `bt_disable()` that triggered HCI activity, so calling `close()` had to be moved after those.

rugeGerritsen replied:
I believe we need https://github.com/zephyrproject-rtos/zephyr/pull/76203
